### PR TITLE
Identity aggregate root changes to allow the "select_raa" FGA option

### PR DESCRIPTION
**Pull request by pablothedude**
This PR will add the changes to the Aggregateroot to allow FGA in middleware. There are however a few changes that still needs to be adressed before this PR could be merged and those changes are:

- [x] Test the accredit identity unhappy flow
- [ ]  RegistrationAuthorityCredentials needs to be done
- [ ]  Auditlog handling needs to be done
- [ ]  Add behat coverage for select_raa (added)
- [ ]  Events have been documented (in class documentation and/or on wiki)

Also the commits need to be rebased before merge to have atomic commits and better commit logs.

See: https://www.pivotaltracker.com/story/show/160283517

### CODE REVIEW

File: src/Surfnet/Stepup/Identity/Api/Identity.php [hunk 1]
```diff
@@ -224,22 +224,24 @@ public function accreditWith(
     );
 
     /**
+     * @param Institution $institution
      * @param RegistrationAuthorityRole $role
      * @return void
      */
-    public function appointAs(RegistrationAuthorityRole $role);
+    public function appointAs(Institution $institution, RegistrationAuthorityRole $role);
```
Comment (MKodde) [inline src/Surfnet/Stepup/Identity/Api/Identity.php]: You could break from the norm and add some additional context to the added parameters in the docblocks. For now I miss insights into why the institution is a requirement for these methods.

File: src/Surfnet/Stepup/Identity/Identity.php [hunk 2]
```diff
@@ -127,9 +128,9 @@ class Identity extends EventSourcedAggregateRoot implements IdentityApi
     private $vettedSecondFactors;
 
     /**
-     * @var RegistrationAuthority
+     * @var RegistrationAuthorityCollection|RegistrationAuthority[]
```
Comment (MKodde) [inline src/Surfnet/Stepup/Identity/Identity.php]: Nitpick, is it actually possible to set an array of RegistratoinAuthority here? Or is the Collection class always used? In that case, I would prefer the removal of `RegistrationAuthority[]`

File: src/Surfnet/StepupMiddleware/ApiBundle/Identity/Repository/RaCandidateRepository.php [hunk 3]
```diff
@@ -53,6 +78,70 @@ public function removeByIdentityId(IdentityId $identityId)
...
+     * @param InstitutionCollection $raInstitutions
+     * @return void
+     */
+    public function removeInstitutionsNotInList(Institution $institution, InstitutionCollection $raInstitutions)
+    {
+        $raCandidates = $this->createQueryBuilder('rac')
+            ->where('rac.raInstitution = :raInstitution')
+            ->andWhere('rac.institution NOT IN (:institutions)')
+            ->setParameter('raInstitution', $institution)
+            ->setParameter('institutions', $raInstitutions)
```
Comment (MKodde) [inline src/Surfnet/StepupMiddleware/ApiBundle/Identity/Repository/RaCandidateRepository.php]: This results in error:
```
"Doctrine\\DBAL\\DBALException: An exception occurred while executing 'SELECT r0_.id AS id_0, r0_.identity_id AS identity_id_1, r0_.ra_institution AS ra_institution_2, r0_.institution AS institution_3, r0_.name_id AS name_id_4, r0_.common_name AS common_name_5, r0_.email AS email_6 FROM ra_candidate r0_ WHERE r0_.ra_institution = ? AND r0_.institution NOT IN (?)' with params [\"institution-a.example.com\", {\"institutions\":[\"institution-a.example.com\",\"institution-d.example.com\"]}]:\n\nCatchable Fatal Error: Object of class Surfnet\\Stepup\\Identity\\Collection\\InstitutionCollection could not be converted to string"
```

Comment (MKodde) [reply src/Surfnet/StepupMiddleware/ApiBundle/Identity/Repository/RaCandidateRepository.php]: Converting to array seems to fix this for me

**Comment by MKodde** [comment]
Do we need a migration, event replay or other mandatory script run to ensure these projections are updated in every installation that upgrades to this version of StepUp? If this is required, do not forget to mention it in the upgrading notes!

File: app/DoctrineMigrations/Version20181023151546.php [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    /**
+     * @param Schema $schema
+     */
+    public function up(Schema $schema)
+    {
+        // this up() migration is auto-generated, please modify it to your needs
+        $this->abortIf($this->connection->getDatabasePlatform()->getName() !== 'mysql', 'Migration can only be executed safely on \'mysql\'.');
+
+        $this->addSql('ALTER TABLE ra_listing DROP PRIMARY KEY');
+        $this->addSql('ALTER TABLE ra_listing ADD id INT AUTO_INCREMENT PRIMARY KEY NOT NULL, ADD ra_institution VARCHAR(255) NOT NULL COMMENT \'(DC2Type:institution)\'');
```
Comment (MKodde) [inline app/DoctrineMigrations/Version20181023151546.php]: Nitpick: maybe position the `id` column at the start of the schema?!

**Comment by MKodde** [comment]
An event replay now fails on an undefined index: ra_institution

```
 Found 204 Events, replaying in increments of 1000
 3/3 [============================] 100%< 1 sec/< 1 sec
 58.5 MiB ERROR OCCURRED: "Notice: Undefined index: ra_institution"
 204/204 [============================] 100%< 1 sec/< 1 sec
 59.2 MiB
In IdentityAccreditedAsRaaEvent.php line 100:
                                           
  Notice: Undefined index: ra_institution  
```

File: src/Surfnet/Stepup/Identity/Api/Identity.php (hunk 1 above)
Comment (pablothedude) [reply src/Surfnet/Stepup/Identity/Api/Identity.php]: Added a doc block

File: src/Surfnet/StepupMiddleware/ApiBundle/Identity/Repository/RaCandidateRepository.php (hunk 3 above)
Comment (pablothedude) [reply src/Surfnet/StepupMiddleware/ApiBundle/Identity/Repository/RaCandidateRepository.php]: Fixed

File: app/DoctrineMigrations/Version20181023151546.php (hunk 4 above)
Comment (pablothedude) [reply app/DoctrineMigrations/Version20181023151546.php]: Fixed

**Comment by pablothedude** [comment]
Add select_ra(a) behat coverage is added as todo

File: src/Surfnet/StepupMiddleware/ApiBundle/Identity/Projector/RaCandidateProjector.php [hunk 5]
```diff
@@ -157,4 +204,62 @@ protected function applyIdentityForgottenEvent(IdentityForgottenEvent $event)
...
+    {
+        $this->raCandidateRepository->removeByRaInstitution(new Institution($event->institution->getInstitution()));
+    }
+
+    /**
+     * @param Institution $institution
+     * @param ConfigurationInstitution[] $authorizedInstitutions
+     * @throws \Doctrine\ORM\NonUniqueResultException
+     */
+    private function updateInstitutionCandidatesFromCollection(Institution $institution, array $authorizedInstitutions)
```
Comment (MKodde) [inline src/Surfnet/StepupMiddleware/ApiBundle/Identity/Projector/RaCandidateProjector.php]: In my dev environment, non ra-candidates seem to end up in the `ra-candidates` projection. The search query later correctly filters these options away so the impact on the read side is not evident. I suspect this behaviour is introduced in this PR, but I'm not 100% sure. Can you verify if this is correct?

File: app/DoctrineMigrations/Version20181022155759.php [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+    /**
+     * @param Schema $schema
+     */
+    public function down(Schema $schema)
+    {
+        // this down() migration is auto-generated, please modify it to your needs
+        $this->abortIf($this->connection->getDatabasePlatform()->getName() !== 'mysql', 'Migration can only be executed safely on \'mysql\'.');
+
+        $this->addSql('ALTER TABLE ra_candidate DROP PRIMARY KEY');
```
Comment (MKodde) [inline app/DoctrineMigrations/Version20181022155759.php]: The `down` part of this migration is not very robust. Once the projection is populated with entries where an Identity is considered RA candidate for multiple insittutions, there are key constraint errors and duplicate entry warnings.

File: src/Surfnet/StepupMiddleware/ApiBundle/Identity/Projector/RaCandidateProjector.php [hunk 7]
```diff
@@ -147,14 +253,57 @@ public function applyRegistrationAuthorityRetractedEvent(RegistrationAuthorityRe
...
+        // loop through authorized institutions
+        foreach ($raInstitutions as $raInstitution) {
+
+            // add new identities
+            $identities = $this->identityRepository->findByInstitution($raInstitution);
+            foreach ($identities as $identity) {
+                $identityId = new IdentityId($identity->id);
+
+                // check if persistent in ra listing
+                if ($this->raListingRepository->findByIdentityIdAndInstitution($identityId, $institution)) {
```
Comment (MKodde) [inline src/Surfnet/StepupMiddleware/ApiBundle/Identity/Projector/RaCandidateProjector.php]: The uniqe constraint is on RA institution and identity id, not on the identity institution. This causes problems when event replaying!

File: src/Surfnet/StepupMiddleware/ApiBundle/Identity/Projector/RaCandidateProjector.php [hunk 8]
```diff
@@ -147,14 +253,57 @@ public function applyRegistrationAuthorityRetractedEvent(RegistrationAuthorityRe
             $event->identityInstitution,
             $event->nameId,
             $event->commonName,
-            $event->email
+            $event->email,
+            $event->raInstitution
```
Comment (MKodde) [inline src/Surfnet/StepupMiddleware/ApiBundle/Identity/Projector/RaCandidateProjector.php]: Is `raInstitution` a property of the RegistrationAuthorityRetractedEvent? In the version I've reviewed, it was not!

File: src/Surfnet/Stepup/Identity/Api/Identity.php (hunk 1 above)
Comment (pablothedude) [reply src/Surfnet/Stepup/Identity/Api/Identity.php]: Added a doc block

File: src/Surfnet/StepupMiddleware/ApiBundle/Identity/Projector/RaCandidateProjector.php (hunk 5 above)
Comment (pablothedude) [reply src/Surfnet/StepupMiddleware/ApiBundle/Identity/Projector/RaCandidateProjector.php]: This behavior was already present and this depends on the token verification logic

File: src/Surfnet/Stepup/Identity/Event/AppointedInstitutionAsRaaEvent.php [hunk 9]
```diff
@@ -0,0 +1,84 @@
+<?php
+
+/**
+ * Copyright 2014 SURFnet bv
```
[the remark this answers is not in the document]
Comment (pablothedude) [reply src/Surfnet/Stepup/Identity/Event/AppointedInstitutionAsRaaEvent.php]: Changed

File: src/Surfnet/Stepup/Identity/Event/AppointedInstitutionAsRaaEvent.php [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+ */
+
+namespace Surfnet\Stepup\Identity\Event;
+
+use Surfnet\Stepup\Identity\AuditLog\Metadata;
+use Surfnet\Stepup\Identity\Value\IdentityId;
+use Surfnet\Stepup\Identity\Value\Institution;
+use Surfnet\Stepup\Identity\Value\NameId;
+
+class AppointedInstitutionAsRaaEvent extends IdentityEvent
```
[the remark this answers is not in the document]
Comment (pablothedude) [reply src/Surfnet/Stepup/Identity/Event/AppointedInstitutionAsRaaEvent.php]: I agree, I chaned the name to be more descriptive

File: src/Surfnet/StepupMiddleware/ApiBundle/Identity/Projector/RaCandidateProjector.php (hunk 7 above)
Comment (pablothedude) [reply src/Surfnet/StepupMiddleware/ApiBundle/Identity/Projector/RaCandidateProjector.php]: I fixed this while working on the event replay.

File: src/Surfnet/StepupMiddleware/ApiBundle/Identity/Projector/RaCandidateProjector.php (hunk 8 above)
Comment (pablothedude) [reply src/Surfnet/StepupMiddleware/ApiBundle/Identity/Projector/RaCandidateProjector.php]: This was fixed while working on the event replay

File: src/Surfnet/Stepup/Identity/Api/Identity.php (hunk 1 above)
Comment (MKodde) [reply src/Surfnet/Stepup/Identity/Api/Identity.php]: There is an error in that doc block:

>  This method will appoint an institution to become ra or raa for another institution

The institution is not appointed right? The Identity is.

Comment (MKodde) [reply src/Surfnet/Stepup/Identity/Api/Identity.php]: That addition does not make that much sense:

> This method will appoint an institution to become ra or raa for another institution

The identity is appointed, not the institution.

File: src/Surfnet/Stepup/Identity/Event/AppointedInstitutionAsRaaEvent.php (hunk 10 above)
[the remark this answers is not in the document]
Comment (MKodde) [reply src/Surfnet/Stepup/Identity/Event/AppointedInstitutionAsRaaEvent.php]: This is an improvement. Like we discussed, I've added a task on this PR to update the event documentation.

**Pull request merged**